language/move-stdlib/src/natives/string.rs

```diff
@@ -51,33 +81,73 @@ pub fn native_is_char_boundary(
         // This is safe because we guarantee the bytes to be utf8.
         std::str::from_utf8_unchecked(s_ref.as_slice()).is_char_boundary(i as usize)
     };
-    // TODO: extensible native cost tables
-    let cost = native_gas(context.cost_table(), NativeCostIndex::PUSH_BACK, 1);
-    NativeResult::map_partial_vm_result_one(cost, Ok(Value::bool(ok)))
+    NativeResult::map_partial_vm_result_one(gas_params.base_cost, Ok(Value::bool(ok)))
+}
+
+pub fn make_native_is_char_boundary(gas_params: IsCharBoundaryGasParameters) -> NativeFunction {
+    Arc::new(
+        move |context, ty_args, args| -> PartialVMResult<NativeResult> {
+            native_is_char_boundary(&gas_params, context, ty_args, args)
+        },
+    )
+}
+
+/***************************************************************************************************
+ * native fun internal_is_sub_string
+ **************************************************************************************************/
+#[derive(Debug, Clone)]
+pub struct SubStringGasParameters {
+    pub base_cost: u64,
+    pub unit_cost: u64,
 }
 
-pub fn native_sub_string(
-    context: &mut NativeContext,
+fn native_sub_string(
+    gas_params: &SubStringGasParameters,
+    _context: &mut NativeContext,
     _ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(args.len() == 3);
     let j = pop_arg!(args, u64) as usize;
     let i = pop_arg!(args, u64) as usize;
+
+    if j < i {
+        // TODO: what abort code should we use here?
+        return Ok(NativeResult::err(gas_params.base_cost, 1));
+    }
+
     let s_arg = pop_arg!(args, VectorRef);
     let s_ref = s_arg.as_bytes_ref();
     let s_str = unsafe {
         // This is safe because we guarantee the bytes to be utf8.
         std::str::from_utf8_unchecked(s_ref.as_slice())
     };
     let v = Value::vector_u8((&s_str[i..j]).as_bytes().iter().cloned());
-    // TODO: extensible native cost tables
-    let cost = native_gas(context.cost_table(), NativeCostIndex::PUSH_BACK, 1);
+
+    let cost = gas_params.base_cost + gas_params.unit_cost * (j - i) as u64;
```

Comment:
The old formula is semantically equivalent to `unit_cost * max(j - i, 1)`. I wonder if it's okay for us to make a change like this, that makes it impossible to maintain fully backward compatibility, no matter the values you pick for your new gas parameters?